File: workspace/2023-01-26/caesar-cipher.haskell/caesar-cipher.hs
import Data.Char
import Text.Printf

main :: IO ()
main = do
    demonstration [1..26] "Hello, World!"

demonstration :: [Int] -> String -> IO ()
demonstration [] s = printf ""
demonstration (n:ns) s = do
    let encoded = encode n s
    let decoded = decode n encoded
    printf "(n = %2d): \"%s\" -> \"%s\" -> \"%s\"\n" n s encoded decoded
    demonstration ns s

encode :: Int -> String -> String
encode n s = map (rotate n) s

decode :: Int -> String -> String
decode n s = encode (26 - n) s

rotate :: Int -> Char -> Char
rotate n c | isLower c = rotateLower n c
           | isUpper c = rotateUpper n c
           | otherwise = rotateNonAlphabet n c
    where
        rotateLower = rotateChar $ ord 'a'
        rotateUpper = rotateChar $ ord 'A'
        rotateNonAlphabet n c = c
        rotateChar offset n c = chr $ (ord c + n - offset) `mod` 26 + offset

File: workspace/2023-01-29/caesar-cipher.ruby/caesar_cipher.rb

def encode(n, s)
    s.chars.map{|c| rotate(n, c)}.join
end

def decode(n, s)
    encode(26 - n, s)
end

def rotate(n, c)
    if lower?(c)
        offset = 'a'.ord
        c = ((c.ord + n - offset) % 26 + offset).chr
    end

    if upper?(c)
        offset = 'A'.ord
        c = ((c.ord + n - offset) % 26 + offset).chr
    end

    c
end

def lower?(c)
    ('a'..'z').include?(c)
end

def upper?(c)
    ('A'..'Z').include?(c)
end

